torch/csrc/jit/type.h

```diff
@@ -51,13 +52,17 @@ struct cloneType<true, T> {
 template<typename T>
 struct cloneType<false, T> {
   std::shared_ptr<T> operator()(std::shared_ptr<const T> ptr) const {
-    return std::make_shared<T>(*ptr);
+    auto result = std::make_shared<typename std::remove_const<T>::type>(*ptr);
+    result->kind_ = T::Kind;
```

Comment:
what is going on here?   Shouldn't the constructor set this?

Reply:
So the problem is that even though we might slice the structure here, copy constructor will preserve `kind_` (as it's a field of `Type`). This means that slicing of `CompleteTensorType` to `TensorType` gives you a struct of the dynamic type, but with `kind_` still set to the old value.

Reply:
I see, maybe add a comment.